src/librbd/migration/HttpClient.cc

```diff
@@ -14,36 +14,162 @@
 #include <boost/asio/read.hpp>
 #include <boost/asio/ssl.hpp>
 #include <boost/beast/core.hpp>
-#include <boost/beast/ssl/ssl_stream.hpp>
+#include <boost/beast/http/read.hpp>
+#include <deque>
 
 namespace librbd {
 namespace migration {
 
-template <typename I>
-struct HttpClient<I>::HttpSessionInterface {
-  virtual ~HttpSessionInterface() {}
-
-  virtual void init(Context* on_finish) = 0;
-  virtual void shut_down(Context* on_finish) = 0;
-};
-
 #define dout_subsys ceph_subsys_rbd
 #undef dout_prefix
 #define dout_prefix *_dout << "librbd::migration::HttpClient::" \
                            << "HttpSession " << this << " " << __func__ \
                            << ": "
 
+/**
+ * boost::beast utilizes non-inheriting template classes for handling plain vs
+ * encrypted TCP streams. Utilize a base-class for handling the majority of the
+ * logic for handling connecting, disconnecting, reseting, and sending requests.
+ */
+
 template <typename I>
 template <typename D>
 class HttpClient<I>::HttpSession : public HttpSessionInterface {
 public:
   void init(Context* on_finish) override {
+    ceph_assert(m_http_client->m_strand.running_in_this_thread());
+
+    auto cct = m_http_client->m_cct;
+    ldout(cct, 15) << dendl;
+
+    ceph_assert(m_state == STATE_UNINITIALIZED);
+    m_state = STATE_CONNECTING;
+
     resolve_host(on_finish);
   }
 
   void shut_down(Context* on_finish) override {
-    disconnect(new LambdaContext([this, on_finish](int r) {
-      handle_shut_down(r, on_finish); }));
+    ceph_assert(m_http_client->m_strand.running_in_this_thread());
+
+    auto cct = m_http_client->m_cct;
+    ldout(cct, 15) << dendl;
+
+    ceph_assert(on_finish != nullptr);
+    ceph_assert(m_on_shutdown == nullptr);
+    m_on_shutdown = on_finish;
+
+    auto current_state = m_state;
+    if (current_state == STATE_UNINITIALIZED) {
+      // never initialized or resolve/connect failed
+      on_finish->complete(0);
+      return;
```

Comment:
should `m_on_shutdown` be reset to `nullptr` here?

Reply:
I avoided clearing `m_on_shutdown` anywhere since `shut_down(...)` should only be called once and there is an assertion on that.